cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java

```diff
@@ -0,0 +1,110 @@
+import io.cdap.cdap.proto.id.ProgramRunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+
+/**
+ * A service that periodically scans for program runs that are in Stopping state and
+ * force terminates them if they are running beyond the expected terminateTime.
+ */
+public class ProgramRunStatusMonitorService extends AbstractRetryableScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunStatusMonitorService.class);
+
+  private final Store store;
+  private final ProgramRuntimeService runtimeService;
+  private final int txBatchSize;
+  private final long intervalMillis;
+
+  @Inject
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService) {
+    this(cConf, store, runtimeService, cConf.getInt(Constants.AppFabric.PROGRAM_TERMINATOR_TX_BATCH_SIZE),
+         cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_SECS));
+  }
+
+  @VisibleForTesting
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize, long specifiedIntervalSecs) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    if (specifiedIntervalSecs <= 0) {
+      LOG.warn("Invalid interval {} specified for the program terminator. Setting it to 5 minutes.",
+               specifiedIntervalSecs);
+      specifiedIntervalSecs = TimeUnit.MINUTES.toMillis(5);
+    }
+    this.intervalMillis = TimeUnit.SECONDS.toMillis(specifiedIntervalSecs);
+  }
+
+  @Override
+  protected long runTask() {
+    terminateStoppingProgramsActiveBeyondTerminateTime();
+    return this.intervalMillis;
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
+    // fetch all runs that are in stopping state that started at most a minute before current time.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    long currentTimeInMillis = System.currentTimeMillis();
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(
+      ProgramRunStatus.STOPPING, 0L, currentTimeInMillis - TimeUnit.MINUTES.toMillis(1), txBatchSize, null);
+    for (Map.Entry<ProgramRunId, RunRecordDetail> stoppingRun : stoppingRuns.entrySet()) {
+      RunRecordDetail runRecordDetail = stoppingRun.getValue();
+      Long terminateTime = runRecordDetail.getTerminateTs();
+      // If terminateTime or RuntimeInfo are null, there isn't anything that can be done about it, so we just ignore it
+      if (terminateTime == null || terminateTime > TimeUnit.MILLISECONDS.toSeconds(currentTimeInMillis)) {
```

Comment:
The current time in seconds is an invariant , should just compute once outside of the loop.

Reply:
currentTime was initialized outside loop but the conversion was being done here. Moved the conversion as well to outside the loop